text/0000-template.md

```diff
@@ -0,0 +1,341 @@
+
+As we cannot trust the peers, the fullnode download the blocks from all peers it is connected to. Let's say we are connected to three peers, the last checkpoint is `b300`, and the best block is `b650`. It means that we will download `[b301, b302, b303, ..., b650]` from all three peers. We will probably receive duplicate blocks and waste some bandwidth.
+
+This phase consists of two sub-steps that run in sequence: (i) discover the common block in the bestchain, and (ii) download blocks.
+
+The discover of the common block is necessary because we might already have some blocks. Let's say the fullnode was synced until block `b402` and it was restarted. In this sub-step, we will find out that `b402` is the common block. The discovery is done through an n-ary search algorithm detailed in the reference level implementation.
+
+After we find out the common block (`b402`), we download blocks from it towards the best block. In our example, we will download blocks `b403`, then `b404`, then `b405`, and so on. After a block is downloaded and verified, the fullnode starts to download its transactions. This example is shown in the figure below:
+
+![fig-phase2-example](./img/fig-phase2-example.svg "Phase 2 Example")
+
+Notice that Phase 1 and Phase 2 can run simultaneously. After both phases and finished, all blocks of the bestchain and their transactions will have been downloaded and verified. It is a benefit for miners, because after phase 2 is finished, the fullnode can start generating block templates for the mining API (even if phase 1 hasn't finished yet). This considerably reduces the required time to start mining (even though their blocks will not verify any transaction before the mempool is synced).
+
+## Phase 3: Sync the mempool
+
+The mempool consists of transactions that have not been verified by any block in the bestchain yet. This phase can only start after phase 1 and phase 2 are finished.
+
+It is critical to have the mempool synced because (i) wallets can quickly show new transactions and update the balance, (ii) miners need the mempool to confirm transactions, and (iii) new transactions should have the most recent transactions in the mempool as parents.
+
+As the transactions form a Direct Acyclic Graph (DAG), the transactions in the mempool form a sub-DAG that connects to the sub-DAG of transactions confirmed by the bestchain (see figure below).
+
+![fig-mempool-example](./img/fig-mempool-example.svg "Mempool Example")
+
+There are several optimizations and custom rules that can be applied in this phase. For example, we can safely discard a transaction that is in conflict with another transaction in the mempool. If it happens that the other transaction was confirmed, it will be downloaded as soon as the new block is synced.
```

Comment:
> we can safely discard a transaction that is in conflict with another transaction in the mempool

Maybe make it clear that we can only safely discard one of them, but not both, or the transaction will be lost.

> If it happens that the ~other~ _discarded_ transaction _is_ confirmed

Using "other" reads to me as the "another" transaction that is in the mempool (that we're keeping), but we're talking about the transaction that we could have discarded.

Reply:
Thanks! Can you review it again, please?

Reply:
Looks good!